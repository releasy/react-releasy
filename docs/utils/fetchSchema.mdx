---
name: fetchSchema
menu: Utils
route: /utils/fetchSchema
---

# fetchSchema

## Usage

```javascript
import { fetchSchema } from 'react-releasy';

fetchSchema({
  url: 'https://yourserveraddress.com/graphql',
  dir: './data',
}).then(() => {
  console.log('Schema downloaded');
  process.exit();
});

```
