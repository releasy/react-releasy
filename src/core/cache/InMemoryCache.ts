import { Variables, GraphQLResponse } from 'relay-runtime';
import RelayQueryResponseCache from 'relay-runtime/lib/RelayQueryResponseCache.js';

import { CacheInterface } from '../Config';

type InMemoryCacheProps = {
  ttl?: number,
  size?: number,
};

class InMemoryCache implements CacheInterface {
  store: RelayQueryResponseCache;

  constructor(props?: InMemoryCacheProps) {
    this.store = new RelayQueryResponseCache({
      size: props && props.size ? props.size : 250,
      ttl: props && props.ttl ? props.ttl : 60 * 1000 * 5, // default = 5 minutes
    });
  }

  get = (queryID: string, variables: Variables): GraphQLResponse => {
    return this.store.get(queryID, variables);
  }

  set = (queryID: string, variables: Variables, payload: GraphQLResponse): void => {
    this.store.set(queryID, variables, payload);
  }

  clear = (): void => {
    this.store.clear();
  }
}

export default InMemoryCache;
