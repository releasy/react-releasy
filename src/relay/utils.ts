import { CacheConfig, RequestNode } from 'relay-runtime';

export const isMutation = (request: RequestNode) => request.operationKind === 'mutation';

export const mustForceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);
