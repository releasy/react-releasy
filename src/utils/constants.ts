export const IS_NODE = typeof window === 'undefined';

export const WINDOW_DATA_KEY = '__RELEASY_DATA__';
