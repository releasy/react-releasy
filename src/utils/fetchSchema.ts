import {
  buildClientSchema,
  printSchema,
  introspectionQuery,
} from 'graphql';
import fse from 'fs-extra';
import fetch from 'node-fetch';

type FetchSchemaInput = {
  url: string,
  path: string,
};

const fetchSchema = async (input: FetchSchemaInput): Promise<void> => {
  const { url, path } = input;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query: introspectionQuery,
    }),
  });

  const result = await response.json();

  const clientSchema = buildClientSchema(result.data);
  const schema = printSchema(clientSchema);

  fse.ensureFile(path);

  return fse.writeFile(path, schema);
};

export default fetchSchema;
