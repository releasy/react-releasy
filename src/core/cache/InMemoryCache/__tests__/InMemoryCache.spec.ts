import { delay } from '../../../../../tests/utils';

import InMemoryCache from '../InMemoryCache';

it('should create a default cache', () => {
  const cache = new InMemoryCache();

  expect(cache).toMatchSnapshot();
});

it('should create a custom cache', () => {
  const cache = new InMemoryCache({
    size: 100,
    ttl: 100,
  });

  expect(cache).toMatchSnapshot();
});

it('should set and get an item', () => {
  const cache = new InMemoryCache();

  const queryID = 'queryID';
  const variables = {};
  const data = { data: 'data' };

  cache.set(queryID, variables, data);

  expect(cache.get(queryID, variables)).toBe(data);
});

it('should clear the cache', () => {
  const cache = new InMemoryCache();

  const queryID = 'queryID';
  const variables = {};
  const data = { data: 'data' };

  cache.set(queryID, variables, data);
  cache.clear();

  expect(cache.store._responses.size).toBe(0);
});

it('should handle size', () => {
  const cache = new InMemoryCache({ size: 1 });

  const firstQueryID = 'firstQueryID';
  const firstVariables = {};
  const firstData = { data: 'firstData' };

  cache.set(firstQueryID, firstVariables, firstData);

  const secondQueryID = 'secondQueryID';
  const secondVariables = {};
  const secondData = { data: 'secondData' };

  cache.set(secondQueryID, secondVariables, secondData);

  expect(cache.store._responses.size).toBe(1);
});

it('should handle ttl', async () => {
  const cache = new InMemoryCache({ ttl: 10 });

  const queryID = 'queryID';
  const variables = {};
  const data = { data: 'data' };

  cache.set(queryID, variables, data);

  expect(cache.get(queryID, variables)).toBe(data);

  await delay(100);

  expect(cache.get(queryID, variables)).toBe(null);
});
