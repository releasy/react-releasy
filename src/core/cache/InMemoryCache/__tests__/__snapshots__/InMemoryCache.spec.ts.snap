// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a custom cache 1`] = `
InMemoryCache {
  "clear": [Function],
  "get": [Function],
  "set": [Function],
  "store": RelayQueryResponseCache {
    "_responses": Map {},
    "_size": 100,
    "_ttl": 100,
  },
}
`;

exports[`should create a default cache 1`] = `
InMemoryCache {
  "clear": [Function],
  "get": [Function],
  "set": [Function],
  "store": RelayQueryResponseCache {
    "_responses": Map {},
    "_size": 250,
    "_ttl": 300000,
  },
}
`;
